packages/driver/src/codecs/record.ts

```diff
@@ -0,0 +1,84 @@
+/*!
+ * This source file is part of the EdgeDB open source project.
+ *
+ * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { ICodec, uuid, CodecKind } from "./ifaces";
+import { Codec } from "./ifaces";
+import { ReadBuffer, WriteBuffer } from "../primitives/buffer";
+import {
+  InvalidArgumentError,
+  ProtocolError,
+} from "../errors";
+
+export class RecordCodec extends Codec implements ICodec {
+  private subCodecs: ICodec[];
+  private names: string[];
+
+  constructor(
+    tid: uuid,
+    codecs: ICodec[],
+    names: string[],
+  ) {
+    super(tid);
+    this.subCodecs = codecs;
+    this.names = names;
+  }
+
+  encode(buf: WriteBuffer, object: any): void {
+    throw new InvalidArgumentError(
+      "SQL records cannot be passed as arguments");
+  }
+
+  decode(buf: ReadBuffer): any {
+    const els = buf.readUInt32();
+    const subCodecs = this.subCodecs;
+    if (els !== subCodecs.length) {
+      throw new ProtocolError(
+        `cannot decode Record: expected ` +
+          `${subCodecs.length} elements, got ${els}`,
+      );
+    }
+
+    const elemBuf = ReadBuffer.alloc();
+    const result: any[] = new Array(els);
+    for (let i = 0; i < els; i++) {
+      buf.discard(4); // reserved
+      const elemLen = buf.readInt32();
+      let val = null;
+      if (elemLen !== -1) {
+        buf.sliceInto(elemBuf, elemLen);
+        val = subCodecs[i].decode(elemBuf);
+        elemBuf.finish();
+      }
+      result[i] = val;
+    }
+
+    return result;
```

Comment:
I thought we were going to decode sql records into some custom datatype where you could access columns by name or index?

Reply:
I'm pesonally fine with this being an array of tuples by default. If we had a custom datatype for this here are some options I've just thought of:

- `asTuples` and `asObjects` methods to make it easy to serialize, but then we pay for the overhead of the custom datatype.
- `toJSON` method that serializes as an array of tuples
- wrap an array of tuples (and has the appropriate getter in the `valueOf` prototype method) and provide a `withNames` method that will wrap it in an object interface with the names

Reply:
I thought so too and then I started digging why we stopped doing custom data types for named tuples and found this: https://github.com/edgedb/edgedb-js/pull/310

tl;dr: we can do custom types but some JS frameworks hardcode their JSON serialization exclusively for Array / Object types and fail with anything that extends them or does anything funky

Reply:
Yeah, I support just making this an array of tuples and worry about exposing a way of getting an array of objects as a separate effort.

Reply:
Yeah, I had forgot about the serialisation issues. Maybe this can be solved with the custom codecs api? We could default to decoding to arrays as it is now, but also provide a builtin 'custom' codec that would decode to objects with names, so users could choose?

Reply:
Yeah, I thought about this too. It might be a slightly different special cased version of `.withCodecs`, because "SQL Record" doesn't really have a type name. Can be symbol though...